pkg/controller/useraccount/useraccount_controller.go

```diff
@@ -213,6 +230,67 @@ func (r *ReconcileUserAccount) ensureIdentity(logger logr.Logger, userAcc *toolc
 	return identity, false, nil
 }
 
+// Sets the finalizers for UserAccount
+func (r *ReconcileUserAccount) setFinalizers(userAcc *toolchainv1alpha1.UserAccount) error {
+	// Add the finalizer if it is not present
+	if !containsString(userAcc.ObjectMeta.Finalizers, userAccFinalizerName) {
+		userAcc.ObjectMeta.Finalizers = append(userAcc.ObjectMeta.Finalizers, userAccFinalizerName)
+		if err := r.client.Update(context.TODO(), userAcc); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// Handles finalizer logic for UserAccount. Returns bool to indicate Requeue
+func (r *ReconcileUserAccount) finalizer(logger logr.Logger, userAcc *toolchainv1alpha1.UserAccount) (bool, error) {
+	logger.Info("deleting UserAccount and subsequent resources", "name", userAcc.Name)
+
+	// Get the Identity associated with the UserAccount
+	identity := &userv1.Identity{}
+	identityName := ToIdentityName(userAcc.Spec.UserID)
+	if err := r.client.Get(context.TODO(), types.NamespacedName{Name: identityName}, identity); err != nil {
+		if errors.IsNotFound(err) {
+			// Request object not found, could have been deleted after reconcile request.
+			// Owned objects are automatically garbage collected. For additional cleanup logic use finalizers.
+			// Return and don't requeue
+			return false, nil
```

Comment:
if `errors.IsNotFound(err)` true then I guess we can continue rather returning.  It's like you want to remove Identity and User as part of finalizer() and you found that Identity is not there (may be already deleted) which is fine and you should continue deleting User.  So we should not return from here.

Reply:
Yeah.. Also I think we don't have to load both identity and user before deleting them. We should do something like:
- Get User. If doesn't exist then proceed. If exists then delete it and exit the reconcile.
- When the user is deleted it will trigger another reconcile. Get Identity. If exists then delete it and exit reconcile.
- When identity is deleted then one more reconcile is triggered. Everything is deleted then so we can delete the finalizer from meta and exit.

Reply:
@alexeykazakov thanks for pointing this out. I have refactored the code to follow this solution instead